File: typescript/src/prompts/confirm.ts
import { formatError, formatQuestion, readLine } from "../terminal.js";
import { type BaseConfig, BasePrompt } from "./base.js";

export interface ConfirmConfig extends BaseConfig<boolean> {
  default?: boolean | null;
}

export class ConfirmPrompt extends BasePrompt<boolean> {
  constructor(config: ConfirmConfig) {
    super({ ...config, default: config.default ?? false });
  }

  get promptType(): string {
    return "confirm";
  }

  protected validateAnswer(value: unknown): boolean {
    // A null answer falls back to the prompt default (which itself defaults to
    // false). The default is normalized to a boolean in the constructor (R5).
    if (value == null) return this.defaultValue ?? false;
    if (typeof value === "boolean") return value;
    if (typeof value === "string") {
      return ["y", "yes", "true", "1"].includes(value.toLowerCase());
    }
    if (typeof value === "number") {
      return Number.isFinite(value) && value !== 0;
    }
    return Boolean(value);
  }

  protected formatAnswer(value: boolean): string {
    return value ? "Yes" : "No";
  }

  protected async executeTerminal(): Promise<boolean> {
    const hint = this.defaultValue ? "Y/n" : "y/N";
    const prompt = formatQuestion(this.message, ` (${hint})`);
    while (true) {
      const result = await readLine(prompt);
      if (!result) return this.defaultValue ?? false;
      const lower = result.trim().toLowerCase();
      if (lower === "y" || lower === "yes") return true;
      if (lower === "n" || lower === "no") return false;
      process.stderr.write(`${formatError("Invalid input. Please enter y or n.")}\n`);
    }
  }
}

File: rust/src/prompts/confirm.rs
use crate::agent::agent_prompt_with_retry;
use crate::errors::Result;
use crate::mode::is_agent_mode;
use crate::terminal::{format_error, format_question, format_success, read_line};
use serde_json::{json, Value};

pub struct ConfirmConfig {
    pub message: String,
    pub default: bool,
}

impl ConfirmConfig {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
            default: false,
        }
    }
}

pub fn confirm(config: ConfirmConfig) -> Result<bool> {
    if is_agent_mode() {
        confirm_agent(&config)
    } else {
        confirm_terminal(&config)
    }
}

pub fn coerce_bool(value: &Value) -> bool {
    match value {
        Value::Bool(b) => *b,
        Value::String(s) => matches!(s.to_lowercase().as_str(), "y" | "yes" | "true" | "1"),
        Value::Number(n) => n.as_f64().unwrap_or(0.0) != 0.0,
        Value::Null => false,
        _ => false,
    }
}

fn confirm_agent(config: &ConfirmConfig) -> Result<bool> {
    let payload = json!({
        "type": "confirm",
        "message": config.message,
        "default": config.default,
    });

    let default = config.default;
    agent_prompt_with_retry(&payload, move |answer| match answer {
        // A null answer falls back to the prompt default (R5).
        Value::Null => Ok(default),
        other => Ok(coerce_bool(&other)),
    })
}

fn confirm_terminal(config: &ConfirmConfig) -> Result<bool> {
    let hint = if config.default { "Y/n" } else { "y/N" };
    let suffix = format!(" ({hint})");
    loop {
        let prompt = format_question(&config.message, &suffix);
        let raw = read_line(&prompt)?;
        if raw.is_empty() {
            eprintln!(
                "{}",
                format_success(&config.message, if config.default { "Yes" } else { "No" })
            );
            return Ok(config.default);
        }
        match raw.trim().to_lowercase().as_str() {
            "y" | "yes" => {
                eprintln!("{}", format_success(&config.message, "Yes"));
                return Ok(true);
            }
            "n" | "no" => {
                eprintln!("{}", format_success(&config.message, "No"));
                return Ok(false);
            }
            _ => {
                eprintln!("{}", format_error("Invalid input. Please enter y or n."));
            }
        }
    }
}
